File: android/app/src/main/kotlin/com/example/video_poc/ZoomVideoBridge.kt
package com.example.video_poc

import android.content.Context
import android.graphics.Color
import android.os.Handler
import android.os.Looper
import android.view.Gravity
import android.view.View
import android.widget.FrameLayout
import android.widget.TextView
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

object ZoomVideoBridge {
    private val mainHandler = Handler(Looper.getMainLooper())
    var eventSink: EventChannel.EventSink? = null

    fun emit(event: String, data: Any? = null) {
        val payload = HashMap<String, Any?>()
        payload["event"] = event
        if (data != null) {
            payload["data"] = data
        }
        mainHandler.post {
            eventSink?.success(payload)
        }
    }
}

class ZoomVideoEventStreamHandler : EventChannel.StreamHandler {
    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        ZoomVideoBridge.eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        ZoomVideoBridge.eventSink = null
    }
}

class ZoomVideoViewFactory(private val isLocal: Boolean) : PlatformViewFactory(StandardMessageCodec.INSTANCE) {
    override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
        return ZoomVideoPlatformView(context, isLocal)
    }
}

class ZoomVideoPlatformView(context: Context, private val isLocal: Boolean) : PlatformView {
    private val container: FrameLayout = FrameLayout(context).apply {
        setBackgroundColor(Color.BLACK)
        val label = TextView(context).apply {
            text = if (isLocal) "Local video" else "Remote video"
            setTextColor(Color.WHITE)
            textSize = 12f
            gravity = Gravity.CENTER
        }
        addView(label, FrameLayout.LayoutParams(
            FrameLayout.LayoutParams.MATCH_PARENT,
            FrameLayout.LayoutParams.MATCH_PARENT
        ))
    }

    override fun getView(): View = container

    override fun dispose() {
        // No-op for placeholder view.
    }
}

File: ios/Runner/ZoomVideoBridge.swift
import Flutter
import UIKit

final class ZoomVideoBridge {
    static let shared = ZoomVideoBridge()
    private init() {}

    var eventSink: FlutterEventSink?

    func emit(event: String, data: Any? = nil) {
        var payload: [String: Any] = ["event": event]
        if let data = data {
            payload["data"] = data
        }
        DispatchQueue.main.async {
            self.eventSink?(payload)
        }
    }
}

final class ZoomVideoEventStreamHandler: NSObject, FlutterStreamHandler {
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        ZoomVideoBridge.shared.eventSink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        ZoomVideoBridge.shared.eventSink = nil
        return nil
    }
}

final class ZoomVideoViewFactory: NSObject, FlutterPlatformViewFactory {
    private let isLocal: Bool

    init(isLocal: Bool) {
        self.isLocal = isLocal
        super.init()
    }

    func create(
        withFrame frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?
    ) -> FlutterPlatformView {
        return ZoomVideoPlatformView(frame: frame, isLocal: isLocal)
    }
}

final class ZoomVideoPlatformView: NSObject, FlutterPlatformView {
    private let container: UIView

    init(frame: CGRect, isLocal: Bool) {
        container = UIView(frame: frame)
        container.backgroundColor = UIColor.black

        let label = UILabel(frame: container.bounds)
        label.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        label.textAlignment = .center
        label.textColor = UIColor.white
        label.font = UIFont.systemFont(ofSize: 12)
        label.text = isLocal ? "Local video" : "Remote video"
        container.addSubview(label)
        super.init()
    }

    func view() -> UIView {
        return container
    }
}
